packages/backend/src/apps/formatter/actions/date-time/transforms/add-subtract-date-time/fields.ts

```diff
@@ -0,0 +1,80 @@
+import { type DurationLikeObject } from 'luxon'
+import z from 'zod'
+
+import { type TransformSpec } from '@/apps/formatter/common/transform-spec'
+import { ensureZodObjectKey } from '@/helpers/zod-utils'
+
+const opTypeEnum = z.enum(['add', 'subtract'])
+
+const timeUnitEnum = z.enum([
+  'seconds',
+  'minutes',
+  'hours',
+  'days',
+  'months',
+  // To simplify date math code, values have the same keys as luxton
+  // duration objects.
+] as const satisfies ReadonlyArray<keyof DurationLikeObject>)
+
+const opsSchema = z.object({
+  opType: opTypeEnum,
+  timeUnit: timeUnitEnum,
+  timeAmount: z
+    .string()
+    .transform((amount) => Number(amount))
+    .pipe(z.number().int().positive()),
+})
+
+export const fieldSchema = z
+  .object({
+    addSubtractDateTimeOps: z.array(opsSchema).min(1),
+  })
+  .transform((params) => ({
+    ops: params.addSubtractDateTimeOps,
+  }))
+
+function sentenceCase(s: string): string {
+  return s.charAt(0).toUpperCase() + s.slice(1)
+}
+
+export const fields: TransformSpec['fields'] = [
+  {
+    label: 'Specify the amount of time you want to add or subtract',
+    key: ensureZodObjectKey(fieldSchema.sourceType(), 'addSubtractDateTimeOps'),
```

Comment:
good stuff sir, is this the standard now hmm

Reply:
haha this is more paranoia from my end, but i love type safety. Can consider using similar techniques if you are keen